### PR TITLE
refactor!: change solutions type of Result class

### LINKED ISSUE #283: Use Particle with Spin projection in solutions of Result
- [ ] change Result to contain solutions of the type `StateTransitionGraph[ParticleWithSpin]` instead of `StateTransitionGraph[dict]`
- [ ] move `initialize_graphs_with_particles()` call into  `__run_csp()` and merge functionality with `__apply_solutions_to_graph()`

### PR DESCRIPTION
**Pull request by spflueger**
Closes #283 

Note that the Result is currently a `Generic` to allow the StateTransitionManager to return the old type of Result, while the solving module is using the new `ParticleWithSpin`.

### CODE REVIEW

File: expertsystem/nested_dicts.py [hunk 1]
```diff
@@ -309,3 +321,85 @@ def _particle_with_spin_projection_to_dict(
         if item["Type"] == "Spin":
             item["Projection"] = spin_projection
     return output
+
+
+def _convert_nodes_to_dict(
```
Comment (redeboer) [inline expertsystem/nested_dicts.py:326]: It doesn't really matter while this function is temporary (right?), but `_convert_nodes_to_dict` modifies its input, while `_convert_edges_to_dict` creates a new object and returns it.

Comment (redeboer) [inline expertsystem/nested_dicts.py:326]: It doesn't really matter while this function is temporary (right?), but `_convert_nodes_to_dict` modifies its input, while `_convert_edges_to_dict` creates a new object and returns it.

File: expertsystem/solving/__init__.py [hunk 2]
```diff
@@ -223,25 +264,33 @@ def _extract_value(props: Dict[Enum, Any], obj_type: Any) -> Any:
 
 
 def _check_arg_requirements(
-    class_type: Type[Any], props: Dict[Enum, Any]
+    class_type: Type[Any], props: Dict[Enum, Any], key_mapping: Optional[dict]
```
Comment (redeboer) [inline expertsystem/solving/__init__.py:267]: How permanent is this `key_mapping`?

File: expertsystem/solving/__init__.py [hunk 3]
```diff
@@ -439,42 +499,105 @@ def _encode_variable_name(variable_info: _VariableInfo, delimiter: str) -> str:
...
 
 
+def _merge_solutions_with_graph(
+    solutions: List[QuantumNumberSolution],
+    graph: StateTransitionGraph[ParticleWithSpin],
+    allowed_particles: ParticleCollection,
+) -> List[StateTransitionGraph[ParticleWithSpin]]:
+    initialized_graphs = []
+
+    logging.debug("merging solutions with graph...")
```
Comment (redeboer) [inline expertsystem/solving/__init__.py:509]: Is this an informative debugging statement?

File: expertsystem/solving/__init__.py [hunk 4]
```diff
@@ -439,42 +499,105 @@
...
     def _create_edge_variables(
         edge_ids: Sequence[int],
-        qn_list: List[
-            Union[
-                ParticleDecayPropertyNames,
-                ParticlePropertyNames,
-                StateQuantumNumberNames,
-            ]
-        ],
+        qn_list: Set[Type[EdgeQuantumNumber]],
```
Comment (redeboer) [inline expertsystem/solving/__init__.py:600]: Yes that's definitely more readible (and safer)

File: expertsystem/solving/__init__.py [hunk 5]
```diff
@@ -439,42 +499,105 @@
...
+    }
+
+    if EdgeQuantumNumbers.isospin_magnitude in state:
+        changes_dict["isospin"] = Spin(
+            state[EdgeQuantumNumbers.isospin_magnitude],
+            state[EdgeQuantumNumbers.isospin_projection],
+        )
+    if EdgeQuantumNumbers.spin_magnitude in state:
+        changes_dict["spin"] = state[EdgeQuantumNumbers.spin_magnitude]
+    return replace(particle, **changes_dict) == particle
```
Comment (redeboer) [inline expertsystem/solving/__init__.py:578]: Interesting, probably also good to use this in the implementation of `create_(anti_)particle`.

File: expertsystem/data.py [hunk 6]
```diff
@@ -149,6 +149,30 @@ class EdgeQuantumNumbers:  # pylint: disable=too-many-instance-attributes
         edge_qn_type.__module__ = "expertsystem.data"
 
 
+# for static typing
+EdgeQuantumNumber = Union[
```
Comment (redeboer) [inline expertsystem/data.py:153]: We may want to split up this module into something like `types` (which is used internally) and `particles` (which is the 'starting point' interface of the `expertsystem`). See also #196 and also the new `amplitude.model` module (#287).

**Review by redeboer — APPROVED** [review]
Good job! I have a few thoughts/questions, but I approve because I know many issues resolve once #254 has been resolved.

File: expertsystem/nested_dicts.py (hunk 1 above)
Comment (spflueger) [reply expertsystem/nested_dicts.py:326]: Yeah, a function with a return would have been a better. But one of the follow up PRs will remove this code again anyway

File: expertsystem/solving/__init__.py (hunk 2 above)
Comment (spflueger) [reply expertsystem/solving/__init__.py:267]: Temporary

File: expertsystem/solving/__init__.py (hunk 5 above)
Comment (spflueger) [reply expertsystem/solving/__init__.py:578]: I thought its used there already, at least I remember I pointed this out before

File: expertsystem/solving/__init__.py (hunk 3 above)
Comment (spflueger) [reply expertsystem/solving/__init__.py:509]: True, I can remove this

File: expertsystem/solving/__init__.py (hunk 5 above)
Comment (redeboer) [reply expertsystem/solving/__init__.py:578]: Can be handled in #294

File: expertsystem/data.py (hunk 6 above)
Comment (spflueger) [reply expertsystem/data.py:153]: Yes, we can address that in #196

File: expertsystem/solving/__init__.py (hunk 3 above)
Comment (spflueger) [reply expertsystem/solving/__init__.py:509]: I will remove it in one of the next PRs as this change is to superficial to rerun the CI for...

**Pull request merged**